eclipse-theia/theia
# opening large or binary files

#### What it does


fix https://github.com/eclipse-theia/theia/issues/3859, fix https://github.com/eclipse-theia/theia/issues/4731

**Important**: the goal of this PR is to prevent a user to crash the browser tab, it does not aim to improve performance in the electron case. See also: https://github.com/eclipse-theia/theia/pull/8152#issuecomment-656710903

- [x] https://github.com/eclipse-theia/theia/pull/7908 has to be landed
- [x] prompt to open a binary file
- [x] prompt to open a large file
- [x] stream content for Monaco editor to avoid blocking the backend and network communications
- [x] limit incremental file update to several megabytes, it will crash the backend for large files, for such files we have to go with streaming

Out of scope:
- electron: bind disk fs provider directly in the renderer process

#### How to test


- try to open binary file, for instance generate a tar file
- try to open/save large file: https://github.com/microsoft/vscode/issues/30180
  - generate medium size ts less than 32 mb, it should be possible to open both in browser and electron
  - generate large size ts more than 32 mb, it should prompt in the browser and can file on load, in electron it should load without the prompt

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
